src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -411,6 +447,149 @@ void ReduceAxesCompute(const nnvm::NodeAttrs& attrs,
+                                  const RType* in_indptr, const IType* in_idx,
+                                  const DType* in_data,
+                                  const int64_t num_rows) {
+    DType sum, residual;
+    mshadow::red::sum::SetInitValue(sum, residual);
+    const IType jval = static_cast<IType>(j);
+    for (RType i = 0; i < num_rows; ++i) {
+      if (in_indptr[i] >= in_indptr[i + 1]) continue;
+      if ((in_idx[in_indptr[i]] <= jval) &&
+          (in_idx[in_indptr[i + 1] - 1] >= jval)) {
+        // Do binary search for j between in_idx[in_indptr[i]] and
+        // in_idx[in_indptr[i+1]]
+        // The assumption here is in_idx for each row is sorted
+        IType start = in_indptr[i];
+        IType end = in_indptr[i + 1] - 1;
+        IType mid;
+        while (start <= end) {
+          mid = (start + end) / 2;
+          if (in_idx[mid] == jval) {
+            mshadow::red::sum::Reduce(sum, in_data[mid], residual);
+            break;
+          } else if (in_idx[mid] < jval) {
+            start = mid + 1;
+          } else {
+            end = mid - 1;
+          }
+        }
+      }
+    }
+    KERNEL_ASSIGN(out_data[j], req, sum);
+  }
+};
+
+template <int req>
+struct SumCsrKernel<req, 1> {
+  template <typename RType, typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data,
+                                  const RType* in_indptr,
+                                  const DType* in_data) {
+    DType sum, residual;
+    mshadow::red::sum::SetInitValue(sum, residual);
+    for (RType k = in_indptr[i]; k < in_indptr[i + 1]; k++) {
+      mshadow::red::sum::Reduce(sum, in_data[k], residual);
+    }
+    KERNEL_ASSIGN(out_data[i], req, sum);
+  }
+};
+
+template <typename xpu>
+void SumCsrImpl(const nnvm::NodeAttrs& attrs, mshadow::Stream<xpu>* s,
+                const NDArray& input, const OpReqType req, NDArray* output) {
+  if (req == kNullOp) return;
+  const ReduceAxesParam& param = nnvm::get<ReduceAxesParam>(attrs.parsed);
+  CHECK_EQ(param.axis.ndim(), 1U) << "sum(csr) only supports axis 0 or 1";
+  CHECK(param.axis[0] == 0 || param.axis[0] == 1)
+      << "sum(csr) only support axis 0 or 1";
+  CHECK(!param.keepdims) << "keepdims not supported for sparse";
+  CHECK(!param.exclude) << "exclude not supported for sparse";
+  int64_t out_data_size = 0;
+  if (param.axis[0] == 0) {
+    out_data_size = input.shape()[1];
+  } else {
+    out_data_size = input.shape()[0];
+  }
+  // only dense output storage type is supported
+  CHECK_EQ(output->storage_type(), kDefaultStorage);
+
+  CHECK_NE(req, kWriteInplace);
+
+  using namespace mshadow;
+  using namespace mxnet_op;
+  using namespace csr;
+  if (req == kWriteTo) {
+    MSHADOW_TYPE_SWITCH(output->data().type_flag_, DType, {
+      Kernel<set_zero, xpu>::Launch(s, out_data_size,
+                                    output->data().dptr<DType>());
+    })
+  }
+
+  if (!input.storage_initialized()) {
```

Comment:
Need to fill zeros for the output here.

Reply:
Good catch! I have fixed it by adding filling zeroes for kWriteTo and kInPlace together.